pkg/handler/iam_handler.go

```diff
@@ -0,0 +1,209 @@
+			nps = append(nps, np)
+		}
+	}
+	return
+}
+
+func (h *IAMHandler) handlePolicy(ctx context.Context, p *v1alpha1.ResourcePolicy, ttl time.Duration) (*v1alpha1.IAMResponse, error) {
+	var iamC iamClient
+	switch strings.Split(p.Resource, "/")[0] {
+	case "organizations":
+		iamC = h.organizationsClient
+	case "folders":
+		iamC = h.foldersClient
+	case "projects":
+		iamC = h.projectsClient
+	default:
+		return nil, fmt.Errorf("unable to match none of the organizations, folders, projects resource")
+	}
+
+	var np *iampb.Policy
+	if err := retry.Do(ctx, h.retry, func(ctx context.Context) error {
+		// Get current IAM policy.
+		cp, err := iamC.GetIamPolicy(
+			ctx,
+			&iampb.GetIamPolicyRequest{Resource: p.Resource})
+		if err != nil {
+			return fmt.Errorf("failed to get IAM policy: %w", err)
+		}
+
+		// Update the policy with new IAM binding additions.
+		updatePolicy(cp, p.Bindings, ttl)
+
+		// Set the new policy.
+		setIamPolicyRequest := &iampb.SetIamPolicyRequest{
+			Resource: p.Resource,
+			Policy:   cp,
+		}
+		np, err = iamC.SetIamPolicy(ctx, setIamPolicyRequest)
+		// Retry when set IAM policy fail.
+		// TODO(#8): Look for specific errors to retry.
+		if err != nil {
+			return retry.RetryableError(fmt.Errorf("failed to set IAM policy: %w, retrying", err))
+		}
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to handle IAM request: %w", err)
+	}
+
+	return &v1alpha1.IAMResponse{Resource: p.Resource, Policy: np}, nil
+}
+
+// Remove expired bindings and add or update new bindings with expiration condition.
+func updatePolicy(p *iampb.Policy, bs []*v1alpha1.Binding, ttl time.Duration) {
+	// Convert new bindings to a role to bindings map.
+	bsMap := convert(bs)
+	// Clean up current policy bindings.
+	var result []*iampb.Binding
+	for _, cb := range p.Bindings {
+		// Skip non-AOD bindings.
+		if cb.Condition == nil || cb.Condition.Title != ConditionTitle {
+			result = append(result, cb)
+			continue
+		}
+		// TODO (#6): Remove expired bindings.
+		// Exclude duplicative Members from current bindings.
+		cb.Members = removeCommonValues(cb.Members, bsMap[cb.Role])
+		if len(cb.Members) > 0 {
+			result = append(result, cb)
+		}
+	}
+	p.Bindings = result
+
+	// Add new bindings with expiration condition.
+	t := time.Now().UTC().Add(ttl).Format(time.RFC3339)
+	for _, b := range bs {
+		newBinding := &iampb.Binding{
+			Condition: &expr.Expr{
+				Title:      ConditionTitle,
+				Expression: fmt.Sprintf("request.time < timestamp('%s')", t),
+			},
+			Members: b.Members,
+			Role:    b.Role,
+		}
+		p.Bindings = append(p.Bindings, newBinding)
+	}
+}
+
+func convert(bs []*v1alpha1.Binding) map[string][]string {
+	m := make(map[string][]string)
+	for _, b := range bs {
+		m[b.Role] = removeCommonValues(m[b.Role], b.Members)
+		m[b.Role] = append(m[b.Role], b.Members...)
+	}
+	return m
+}
+
+// Returns a result list of strings in l1 that are not found in l2.
+func removeCommonValues(l1, l2 []string) []string {
```

Comment:
Why not use a map? `map[string]struct{}`

Reply:
Could you provide more info?

Reply:
IIUC, the reason you need to remove common values is just to deduplicate? And you do that with a map

Reply:
I see, so map[string]struct{} is a set in golang. Sure, updating.

Reply:
I was wonder whether it'd be easier if `toBindingsMap` returns `map[string]map[string]struct{}`. You don't need to "removeCommonValues" if you're using a map, right?

Reply:
removeCommonValues is also used in exclude common Members from current bindings found in the new bindings, see line 158. Is it still worth to change toBindingsMap?